### ISSUE TITLE
Swappability of main models

### ISSUE THREAD

**Issue by skirsdeda** [wagtail/wagtail#3282]
So I have this problem with Wagtail that sometimes (or more like, usually) main models (wagtailcore.Page, wagtailimages.Image, etc.) need to have some additional functionality or overrides to existing functionality. 

Consider for example making Page translatable without duplicating trees. One would use something like django-modeltranslation to do that but then all the slug/url_path related functions have to be overriden to handle translated urls properly. You could of course write a Mixin and extend all your Page extending models with that, but some of Page's core functionality doesn't necessarily use ```instance.specific``` which means that the only way to override Page's functionality is monkey-patching. And everyone knows how bad monkey-patching is :)

Django has this very underused feature called swappable models (as used in django.contrib.auth.User). It would be great if Wagtail's core models would be swappable. This would be as simple as moving all implementation to abstract base models (like what is already the case with Image model) and leaving actual model empty with only swappable option in it's Meta. Like this:
```python
class Page(BasePage):
    class Meta(BasePage.Meta):
        swappable = 'WAGTAIL_PAGE_MODEL'
```
Then implement a simple function ```get_page_model``` which returns the actual Page model and replace all cases of direct ```Page``` class usage with ```get_page_model()```.

I have attempted a similar refactoring in django-filer and there are really only pros and no cons with this approach. It would all be 100% backwards-compatible. On the other hand, a lot of flexibility is gained this way.

Now I would be willing to provide a pull request for this, but first I wanted to check whether core developers would support such an idea.

**Comment by gasman** [comment]
Hi @skirsdeda,
Yes, this would be very welcome! The lack of a good way to extend the base Page model has been a long-running bugbear (discussed for example here https://groups.google.com/d/msg/wagtail/4459qj1tNiU/D91COykMSmcJ and filed as #836).

Up to now I've been wary of going ahead with this change, because swappable image/document models have been quite a troublesome thing to maintain (for example, see #2881 where we've had to phase in a schema change across 3-4 releases to ensure that each migration is simple enough to be handled by Django's migration autodetector)... intuitively it seems that Page would throw up even more problems, being more complex and more central to Wagtail. I'm very encouraged to know that you've been able to make it work on django-filer, and I look forward to seeing your pull request!

**Comment by skirsdeda** [comment]
@gasman,
Thanks for feedback. I'll look into issues you mentioned a bit deeper. From what I have seen so far in other projects, a lot of painful situations with migrations arise when trying to swap out models without using **swappable** option in Meta and not using **swappable_dependency** in migrations themselves. Then of course, you have to make sure you don't use the swappable model class directly (especially when defining foreign keys to it). Anyway, let's see how this goes...

**Comment by chimeno** [comment]
This issue lead me to https://github.com/wq/django-swappable-models , it's a concept I didn't  know until know, so thanks, and I put the link above in case it helps.

**Comment by mikedingjan** [comment]
@skirsdeda Did you by change already had some time to look into this?

Perhaps I can try to help with this? For our multi language module ``wagtailtrans`` this would really be a welcome functionality.

**Comment by skirsdeda** [comment]
@mikedingjan Not yet, unfortunately. But it should be pretty easy to get a working proof of concept. I mean most of it will be pretty mechanical work of replacing `Page` imports with `get_page_model()`.

**Comment by mikedingjan** [comment]
True, however, providing the user with a good abstract ``BasePage`` which can be used to implement a custom page, a bit like django's AbstractBaseUser class for example.

**Comment by dmarcelino** [comment]
This would be a great help to make the implementation of `wagtail-modeltranslation` cleaner and more robust. I've been trying to improve `wagtail-modeltranslation` and now I've reached a moment where I *need* to make changes to `Page` (https://github.com/infoportugal/wagtail-modeltranslation/pull/150). With this feature things would be much simpler.

**Comment by dfirst** [comment]
Hi everyone. I want to propose an idea of how to make wagtail apps more flexible. Here is raw proof of concept with tests passes:
https://github.com/dfirst/wagtail/tree/PROPOSE_forking_wagtail_apps
Short summary of idea:
1. Use dynamic imports of classes we want to be customizable.
2. For customization i used idea of locale fork of wagtail and dynamic importer. First importer check if class exists in fork if not then it import it from root.(fork path can be controlled WAGTAIL_FORK)
3. If user want to customize a model then he should copy migrations and set MIGRATION_MODULES.
4. The idea can be implemented without breaking changes step by step.
What do you think about it? If the idea is interesting we can create a plan and i will finish it.

**Comment by sonnybaker** [comment]
Is there still interest in moving this forward? @gasman I'd be really keen to hear your thoughts.

**Comment by dmarcelino** [comment]
@sonnybaker, from `wagtail-modeltranslation`'s perspective a swappable `Page` model would be a game changer. It would make the implementation cleaner without recurring to monkey patching `Page`.

cc: @DiogoMarques29

**Comment by gasman** [comment]
@sonnybaker I'd be very keen to see this happen, but I'm too busy to work on it myself. Based on the experience of swappable images/docs, I expect it'll create extra maintenance work in future when dealing with migrations, but I think that's an acceptable tradeoff for the benefits (and perhaps the Page model is pretty stable by now and unlikely to need future migrations anyhow).

**Comment by m5seppal** [comment]
Any ongoing activity with this feature? This would be very much needed. Most of the issues I'm having with the Wagtail at the moment comes from the fact that there is no way to extend the Page model.

**Comment by thibaudcolas** [comment]
Converting this to Discussions. I believe this is happening / has happened, albeit very slowly. There’s [RFC 53: Swappable Page model](https://github.com/wagtail/rfcs/pull/53) that badly needs reviews.
